Use ObservationBox in events

nilsdeppe opened the pull request:
## Proposed changes

Only the commit `Use ObservationBox in Events` is new. Uses the `ObservationBox` in events, switching over the `EventsAndTriggers` classes as well as individual events.

Depends on:
- [x] #3556

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Use ObservationBox in Events

nilsvu commented on src/Evolution/EventsAndDenseTriggers/EventsAndDenseTriggers.hpp:
```diff
@@ -238,6 +244,12 @@ void EventsAndDenseTriggers::run_events(
+          tmpl::at<typename Metavariables::factory_creation::factory_classes,
+                   Event>,
+          get_tags<tmpl::_1>>>,
+      db::is_compute_tag<tmpl::_1>>>;
```
I suggest you template the `EventsAnd(Dense)Triggers` class and the `RunEventsAndTriggers` action on `ComputeTagsForObservations`. Then you can drop most (all?) of #3558 that adds the template parameter to individual events. It makes events easier to add and to maintain. They don't need to know _how_ the tags get into the box, but they only retrieve them. The executable is in control of adding the compute tags either to the DataBox, or to the ObservationBox in `RunEventsAndTriggers`.

nilsdeppe replied:
This was discussed with @wthrowe and @kidder and the preference was this implementation. You need to template the events on the tags they need anyway, either the compute or simple tags. For example, `ObserveFields` could be templated on `EnergyDensityCompute` instead of `EnergyDensity`. This means that each event lists what compute tags it wants (avoids giant magic collections and inferring info from up the tree). With the change you suggested you need to keep the simple tags that are passed as template parameters to the events in sync with the compute tags that you specify to the actions. I think the implementation here is actually much clearer and easier to maintain, which was the consensus with @wthrowe and @kidder

nilsvu replied:
Here's a very specific problem I see with the current implementation, and that would be solved by the idea "the code that retrieves tags out of the box doesn't need to know how the tag got in there":

See https://github.com/sxs-collaboration/spectre/blob/develop/src/Elliptic/Executables/Xcts/SolveXcts.hpp#L258.
The compute item `Xcts::Tags::SpacetimeQuantities` takes a list of possible `gr::Tags` that it is able to compute as template parameters, and returns them in a `Variables`, e.g. the spatial metric, lapse, shift and Hamiltonian constraint. In the current implementation I'd have to pass the `Xcts::Tags::SpacetimeQuantities` compute item to every event, making sure I pass the same instantiation to all to avoid conflicts. Then each event may retrieve a subset of the `gr::Tags`. It would be much cleaner to pass `Xcts::Tags::SpacetimeQuantities` to `RunEventsAndTriggers` to make the subitems available to all events, and then each event may retrieve what it needs.

nilsdeppe replied:
I don't understand what you mean by "same instantiation" here so I can't really comment on that. What you are calling a drawback I'm calling a feature. You _should_ have to specify what the event needs access to. I find it much more difficult to have to figure out why or if something was added if it's added in a completely different location for something like observing. Some parts of the evolution code already having actions like "InitializeConstraints" that add partial derivatives, pointwise L2 norms, and 3+1 items into the DataBox. This makes it _really_ difficult to figure out what you actually need to add to observe something. My view is that it is the events job to ensure what it will need is in the box, which is exactly how the current design works. This is exactly what we do with global cache quantities currently. I'll also note that "every Event" -> "every Event that needs it," which so far is ~2.

nilsvu replied:
Hm that sort of dependency resolution is not how we currently do things.. All our actions get<> things out of the DataBox assuming it was put in there by initialization code. Even with your current implementation there are plenty of things the event retrieves but doesn't add to the box itself, like coordinates, mesh, time.

I agree the event should specify what it needs access to, and it does since it's templated on the tags that it retrieves and specifies `argument_tags`. What I'm saying is that the event shouldn't be responsible for how those tags get _into_ the box.

With 'same instantiation' I mean that both the event that wants to observe constraint norms, as well as the event that observes lapse, shift etc in the volume, need to be templated on the same `Xcts::Tags::SpacetimeQuantities` compute tag that adds both the constraint norms _and_ the lapse, shift etc to the observation box. This goes against the idea that the an event only adds what it needs.

I also don't love the proliferation of a magic type alias like `compute_tags_for_observation_box` that is read from some other place in the code. At the very least it needs to be documented in the `Event` class, so when a person implements a new event they know what interface is expected from them.

nilsdeppe replied:
We can talk about this on Monday with @wthrowe and @kidder.

nilsvu commented on src/ParallelAlgorithms/EventsAndTriggers/EventsAndTriggers.hpp:
```diff
@@ -40,12 +50,24 @@ class EventsAndTriggers {
+                     Event>,
+            get_tags<tmpl::_1>>>,
+        db::is_compute_tag<tmpl::_1>>>;
+    std::optional<ObservationBox<db::detail::expand_subitems<compute_tags>,
```
Perhaps get the type with `decltype()`? We probably shouldn't use `db::detail` here..

nilsdeppe replied:
sorry, I'm not sure how `decltype()` is supposed to replace `expand_subitems`. Could you clarify? I can also just move `expand_subitems` out of the `detail` namespace :shrug:

nilsvu replied:
I was thinking maybe something like `decltype(make_observation_box<compute_tags>(declval<DataBoxType>()))` could work

wthrowe replied:
I like @nilsleiffischer's suggestion, but moving `expand_subitems` is also fine with me.

nilsdeppe commented:
I rebased to remove `noexcept`s and `ElementLogical` frame things, but have not included any changes for the comments since I'm waiting on replies for those. Just so you know what changed :)

wthrowe commented on src/Evolution/EventsAndDenseTriggers/EventsAndDenseTriggers.hpp:
```diff
@@ -234,6 +240,12 @@ void EventsAndDenseTriggers::run_events(
+          tmpl::at<typename Metavariables::factory_creation::factory_classes,
+                   Event>,
+          get_tags<tmpl::_1>>>,
+      db::is_compute_tag<tmpl::_1>>>;
```
Should we just say all things in `compute_tags_for_observation_box` must be compute tags and then the filter is unnecessary?  (Again for plain EaT.)

nilsdeppe replied:
We could, but then I think each individual event needs to have a filter. The reason being that for example ObserveFields has as a template parameter all tags it wants to observe, both simple and compute. Requiring them to all be compute tags would mean we have to then check which tags are already in the box (evolved vars, etc.) and omit those. I guess each event could explicitly have as template parameters the simple tags, compute tags, and vars compute tags. The reason I didn't do this is because it's quite a bit more refactoring. We could require all the tags be compute tags and then just filter based on whether the simple (base) tag of the compute tag is retrievable. That still will require possibly writing compute tags for evolved variables, etc. or some other major refactor of individual events.

wthrowe replied:
OK.  This is fine, then.

nilsdeppe commented:
I've:
- rebased
- switched to `decltype()`
- added an empty `compute_tags_for...` to `control_system::Event`
- squashed the above two small changes

Thanks for the reviews!

nilsvu commented on src/ControlSystem/Event.hpp:
```diff
@@ -78,6 +78,8 @@ class Event : public ::Event {
   static constexpr bool factory_creatable = false;
   Event() = default;
 
+  using compute_tags_for_observation_box = tmpl::list<>;
```
Could you add a bit of documentation either here, or to the class docs, so a person implementing a subclass knows they are supposed to add this alias, and they have an idea what it is

nilsdeppe commented:
I've added documentation to the `Event` base class. Thanks for the reviews!
